### PR TITLE
[Dashboard] Add tune API to support tune tab in new backend

### PR DESCRIPTION
**Pull request by mfitton**
## Why are these changes needed?
These changes are needed to enable running the existing Tune tab with the new backend. There are a few changes to the tune backend to make it work with async io, but otherwise it is pretty much identical to how it was in the old dashboard backend.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

I tested this alongside many other changes in a branch that contains the full migration code. That said, this is not yet unit tested.

### CODE REVIEW

**Comment by simon-mo** [comment]
cc @richardliaw

File: dashboard/modules/tune/tune_head.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            self._logdir = os.path.expanduser(experiment)
+            return None, {"experiment": self._logdir}
+        else:
+            return "Not a Valid Directory", None
+
+    def _enable_tensorboard(self):
+        if not self._tensor_board_dir:
+            tb = program.TensorBoard()
+            tb.configure(argv=[None, "--logdir", str(self._logdir)])
+            tb.launch()
```
Comment (richardliaw) [inline dashboard/modules/tune/tune_head.py:98]: does this eventually get terminated?

Comment (mfitton) [reply dashboard/modules/tune/tune_head.py:98]: @richardliaw I just ran it and tensorboard died when I killed my ray cluster. It seems to fate share with its parent process.

File: dashboard/modules/tune/tune_head.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+import ray.new_dashboard.modules.tune.tune_consts \
+    as tune_consts
+import ray.new_dashboard.utils as dashboard_utils
+from ray.new_dashboard.utils import async_loop_forever, rest_response
+
+try:
+    from ray.tune import Analysis
+    from tensorboard import program
+except ImportError:
+    Analysis = None
```
Comment (fyrestone) [inline dashboard/modules/tune/tune_head.py:16]: What happened if the `Analysis` is None?

File: dashboard/modules/tune/tune_head.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+class TuneController(dashboard_utils.DashboardHeadModule):
+    def __init__(self, dashboard_head):
+        """
+        This dashboard module is responsible for enabling the Tune tab of
+        the dashboard. To do so, it periodically scrapes Tune output logs,
+        transforms them, and serves them up over an API.
+
+        Args
+            logdir (str): Directory path to save the status information of
```
Comment (fyrestone) [inline dashboard/modules/tune/tune_head.py:30]: The `Args` comments can be removed? Or, can we move the comments to `self._logdir`?

File: dashboard/modules/tune/tune_head.py (hunk 1 above)
Comment (richardliaw) [reply dashboard/modules/tune/tune_head.py:98]: sgtm then

File: dashboard/modules/tune/tune_head.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            success=True,
+            message="Fetched tune availability",
+            result=availability)
+
+    @routes.get("/tune/set_experiment")
+    async def set_tune_experiment(self, req) -> aiohttp.web.Response:
+        try:
+            experiment = req.query["experiment"]
+        except KeyError:
+            return await rest_response(success=False, message="Bad request")
```
Comment (fyrestone) [inline dashboard/modules/tune/tune_head.py:60]: The `@routes.get` decorator catches the handler exception and returns `rest_response(success=False, message=<exception stack>)`. This `try ... except ...` do the same thing, but makes the response much cleaner `rest_response(success=False, message="Bad request")`.

File: dashboard/modules/tune/tune_head.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+    def _enable_tensorboard(self):
+        if not self._tensor_board_dir:
+            tb = program.TensorBoard()
+            tb.configure(argv=[None, "--logdir", str(self._logdir)])
+            tb.launch()
+            self._tensor_board_dir = self._logdir
+
+    def _get_availability(self):
+        return {"available": True, "trials_available": self._trials_available}
```
Comment (fyrestone) [inline dashboard/modules/tune/tune_head.py:102]: It seems that the `_get_availability` is only used by the `get_availability`, is it better to merge them?

File: dashboard/modules/tune/tune_head.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+                        "job_id": os.path.basename(self._logdir),
+                        "trial_id": "No Trial ID"
+                    }
+                    other_data = df[df["logdir"].str.contains(trial)]
+                    if len(other_data) > 0:
+                        trial_id = other_data["trial_id"].values[0]
+                        self._errors[str(trial)]["trial_id"] = str(trial_id)
+                        if str(trial_id) in self._trial_records.keys():
+                            self._trial_records[str(trial_id)]["error"] = text
+                            self._trial_records[str(trial_id)][
```
Comment (fyrestone) [inline dashboard/modules/tune/tune_head.py:125]: There are too many `str(trail_id)`, can we just `trail_id = str(other_data["trial_id"].values[0])`?

File: dashboard/modules/tune/tune_head.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+        # list of static attributes for trial
+        default_names = [
+            "logdir", "time_this_iter_s", "done", "episodes_total",
+            "training_iteration", "timestamp", "timesteps_total",
+            "experiment_id", "date", "timestamp", "time_total_s", "pid",
+            "hostname", "node_ip", "time_since_restore",
+            "timesteps_since_restore", "iterations_since_restore",
+            "experiment_tag", "trial_id"
+        ]
```
Comment (fyrestone) [inline dashboard/modules/tune/tune_head.py:185]: If we don't care the order of the names, can we change the `default_names` to a set?
```
timeit.timeit("'iterations_since_restore' in default_names_set", globals=globals())
0.04095575900282711
timeit.timeit("'iterations_since_restore' in default_names", globals=globals())
0.2835074480099138
```

File: dashboard/modules/tune/tune_head.py (hunk 2 above)
Comment (mfitton) [reply dashboard/modules/tune/tune_head.py:16]: If Analysis is none, the Tune tab availability check will fail, and the tune tab won't be displayed.

File: dashboard/modules/tune/tune_head.py (hunk 3 above)
Comment (mfitton) [reply dashboard/modules/tune/tune_head.py:30]: Good point, thank you.

File: dashboard/modules/tune/tune_head.py (hunk 2 above)
Comment (fyrestone) [reply dashboard/modules/tune/tune_head.py:16]: Thanks. The `self.collect()` does not raise an exception? That will be good.

Comment (mfitton) [reply dashboard/modules/tune/tune_head.py:16]: It will not, I added a check.

**Comment by mfitton** [comment]
@fyrestone @richardliaw could you please take a look to see if your concerns are addressed and approve if so?

**Comment by richardliaw** [comment]
Great - could you post a screenshot of this working?

**Comment by mfitton** [comment]
@richardliaw This PR doesn't make this implementation the default dashboard yet. Currently I'm shepherding through multiple PRs to be able to make a PR to switch over to this dashboard by default. 

Given that, this branch doesn't even have the front-end changes included for me to show you this working. I did test it on a branch where I had integrated all these changes before making the PRs.

When I make a PR to deprecate the old dashboard and turn this functionality on, I'll tag you in it and provide a screenshot. Is that alright @richardliaw ?

**Comment by richardliaw** [comment]
Sounds good! No objections for merging from me.

**Comment by mfitton** [comment]
I'm not authorized to merge. @rkooo567 could you please merge?

**Review by rkooo567 — APPROVED** [review]

**Pull request merged**